hypothesis-python/src/hypothesis/internal/conjecture/choice.py

```diff
@@ -0,0 +1,318 @@
+    # obviously-too-large. callers will generally turn this into a .mark_overrun().
+    if size >= BUFFER_SIZE_IR:
+        raise ChoiceTooLarge
+
+    # subtract out the amount responsible for the size
+    index -= _size_to_index(size, alphabet_size=alphabet_size)
+    vals = []
+    for i in reversed(range(size)):
+        # optimization for common case when we hit index 0. Exponentiation
+        # on large integers is expensive!
+        if index == 0:
+            n = 0
+        else:
+            n = index // (alphabet_size**i)
+            # subtract out the nearest multiple of alphabet_size**i
+            index -= n * (alphabet_size**i)
+        vals.append(from_order(n))
+    return vals
+
+
+def zigzag_index(value, *, shrink_towards):
+    # value | 0  1 -1  2 -2  3 -3  4
+    # index | 0  1  2  3  4  5  6  7
+    index = 2 * abs(shrink_towards - value)
+    if value > shrink_towards:
+        index -= 1
+    return index
+
+
+def zigzag_value(index, *, shrink_towards):
+    assert index >= 0
+    # count how many "steps" away from shrink_towards we are.
+    n = (index + 1) // 2
+    # now check if we're stepping up or down from shrink_towards.
+    if (index % 2) == 0:
+        n *= -1
+    return shrink_towards + n
+
+
+def choice_to_index(choice, kwargs):
+    # This function takes a choice in the choice sequence and returns the
+    # complexity index of that choice from among its possible values, where 0
+    # is the simplest.
+    #
+    # Note that the index of a choice depends on its kwargs. The simplest value
+    # (at index 0) for {"min_value": None, "max_value": None} is 0, while for
+    # {"min_value": 1, "max_value": None} the simplest value is 1.
+    #
+    # choice_from_index inverts this function. An invariant on both functions is
+    # that they must be injective. Unfortunately, floats do not currently respect
+    # this. That's not *good*, but nothing has blown up - yet. And ordering
+    # floats in a sane manner is quite hard, so I've left it for another day.
+
+    if isinstance(choice, int) and not isinstance(choice, bool):
+        # Let a = shrink_towards.
+        # * Unbounded: Ordered by (|a - x|, sgn(a - x)). Think of a zigzag.
+        #   [a, a + 1, a - 1, a + 2, a - 2, ...]
+        # * Semi-bounded: Same as unbounded except stop on one side when you hit
+        #   {min, max}_value. so min_value=-1 a=0 has order
+        #   [0, 1, -1, 2, 3, 4, ...]
+        # * Bounded: Ordered by (sgn(a - x), |a - x|). Count upwards until max_value,
+        #   then count downards.
+        #   [a, a + 1, a + 2, ..., max_value, a - 1, a - 2, ..., min_value]
+        #
+        # To simplify and gain intuition about this ordering, you can think about
+        # the most common case where 0 is first (a = 0). We deviate from this only
+        # rarely, e.g. for datetimes, where we generally want year 2000 to be
+        # simpler than year 0.
```

Comment:
I think we want to use the same comparison order for bounded intervals as for unbounded or semi-bounded.

Reply:
Thanks for catching this, have done so. I think I was trying to match the existing ordering at the time (for no principled reason).